### ISSUE TITLE
musl stack size

### ISSUE THREAD

**Issue by wangp** [ivmai/bdwgc#159]
Hi Ivan, I saw you were looking into the small stack size issue with musl. I managed to get gctest to pass with --enable-parallel-marking --enable-gc-assertions after the following hacks. A few other tests fall afoul of the minimum stack size assertion in pthread_support.c (line 1788) because they also create threads with the default attributes.

Hope this helps.

~~~diff
diff --git a/pthread_support.c b/pthread_support.c
index ee4c599..05fd0bd 100644
--- a/pthread_support.c
+++ b/pthread_support.c
@@ -410,7 +410,7 @@ GC_INNER void GC_start_mark_threads_inner(void)
     if (0 != pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_DETACHED))
         ABORT("pthread_attr_setdetachstate failed");
 
-#   if defined(HPUX) || defined(GC_DGUX386_THREADS)
+#   if 1 /* musl */ || defined(HPUX) || defined(GC_DGUX386_THREADS)
       /* Default stack size is usually too small: fix it. */
       /* Otherwise marker threads or GC may run out of    */
       /* space.                                           */
diff --git a/tests/test.c b/tests/test.c
index d6a091b..21a1d6a 100644
--- a/tests/test.c
+++ b/tests/test.c
@@ -535,9 +535,16 @@ void check_marks_int_list(sexpr x)
 
     void fork_a_thread(void)
     {
+      pthread_attr_t attr;
       pthread_t t;
       int code;
-      if ((code = pthread_create(&t, 0, tiny_reverse_test, 0)) != 0) {
+
+      /* Hack for musl. */
+#     define MIN_STACK_SIZE (8*HBLKSIZE*sizeof(word))
+      pthread_attr_init(&attr);
+      pthread_attr_setstacksize(&attr, MIN_STACK_SIZE);
+
+      if ((code = pthread_create(&t, &attr, tiny_reverse_test, 0)) != 0) {
         GC_printf("Small thread creation failed %d\n", code);
         FAIL;
       }
@@ -2170,7 +2177,7 @@ int main(void)
       GC_printf("pthread_attr_init failed, error=%d\n", code);
       FAIL;
     }
-#   if defined(GC_IRIX_THREADS) || defined(GC_FREEBSD_THREADS) \
+#   if 1 /* musl */ || defined(GC_IRIX_THREADS) || defined(GC_FREEBSD_THREADS) \
         || defined(GC_DARWIN_THREADS) || defined(GC_AIX_THREADS) \
         || defined(GC_OPENBSD_THREADS)
         if ((code = pthread_attr_setstacksize(&attr, 1000 * 1024)) != 0) {
~~~

**Comment by ivmai** [comment]
OK. I will prepare the patch.
Related issue: #73

**Comment by ivmai** [comment]
Note for me: remove --disable-parallel-mark for compiler:musl-gcc after the fix.

**Comment by ivmai** [comment]
I have applied your patch (with some refactoring) to master: f9d1419

A better solution could be to adjust stack size in GC_pthread_create (but we should not modify client's attr object).

**Comment by wangp** [comment]
Thanks!

It doesn't seem like there's a function to copy pthread_attr_t.

**Comment by ivmai** [comment]
Correct, we cannot copy it. Probably it would not be a big issue if we adjust stack size of the client's attribute in pthread_create and revert it  back before return from pthread_create.
Anyway, I can't reproduce failures of other tests (with parallel marker on).

**Comment by ivmai** [comment]
BTW. I found another issue with musl+parallel_mark: #169

**Comment by ivmai** [comment]
To @hboehm : Why do we put local_mark_stack to stack instead of using a preallocated memory (e.g. using scratch_alloc)? Some targets have threads' small stack by default.

void GC_do_parallel_mark(void) {
    mse local_mark_stack[LOCAL_MARK_STACK_SIZE];
                /* Note: local_mark_stack is quite big (up to 128 KiB). */
...
void GC_help_marker(word my_mark_no) {
    unsigned my_id;
    mse local_mark_stack[LOCAL_MARK_STACK_SIZE];
...
int WRAP_FUNC(pthread_create)(pthread_t *new_thread, ...) {
    ....
    /* We resist the temptation to muck with the stack size here,       */
    /* even if the default is unreasonably small.  That's the client's  responsibility. */

**Comment by hboehm** [comment]
I think that was a mistake to stack allocate it in GC_do_parallel_mark.
It's probably usually fine in GC_help_marker, since that uses dedicated
threads.

If we allocate it elsewhere, we need to make sure that it's not
accidentally scanned by the collector. scratch_alloc should be OK.

**Comment by ivmai** [comment]
To @wangp : I have moved local mark stack to heap, so clients no longer need to increase threads' stack. Please check the latest master with musl - all tests should pass.

**Comment by wangp** [comment]
That's great. The tests pass for me, too.

**Comment by ivmai** [comment]
> I have moved local mark stack to heap, so clients no longer need to increase threads' stack

Just a note for me: as a consequence the above fix reverts commit 02ae64481 in particular (#202). Thus, e.g. downstream https://git.alpinelinux.org/aports/tree/main/gc/0001-Fix-gctest-with-musl-libc-on-s390x.patch is no longer needed (in v8.2.4 at least).
